### PR TITLE
update intercompatible scripts and add a demo notebook

### PR DESCRIPTION
**Pull request by erialC-P**
This PR updates the data producing intertidal modules and makes them intercompatible, working from a single master ds variable throughout the workflow. Suitable arrays have been converted as integer outputs. A notebook that runs the whole workflow is added - note that some modules are dependent on others being run first. The optimal ordering of modules is recommended in the demo notebook.

### CODE REVIEW

**Review by robbibt — COMMENTED** [review]
Hey @erialC-P, this is super awesome - so exciting to see all the different parts come together! Code looks great - I've left a bunch of small comments on the code but hopefully they're all pretty minor and easy to update. 

More generally, I did have one larger comment: I'm not 100% sure about the current approach of using a single `xr.Dataset` to hold all of the temporary outputs and pass them between the various functions. A few thoughts on this:

* By having the main input to each function be the single `ds`, we make the code opaque and difficult for users to anticipate exactly what inputs each function needs. For example, `ds` can contain very different layers at different points in the workflow, so passing the same dataset in at different times could produce very different results (and make it tricky to re-create the one large layer exactly as the function requires). 
* By passing `ds` in and out of the various functions, we risk accidently making copies of the data in memory. My grasp of how Python deals with that stuff is hazy, but I know that often weird things happen when data is assigned to a dataset within a function's scope vs. outside of the function call
* By filling up `ds` with a mix of 2D and 3D arrays with non-matching dimensions (like `tide_cq`, `diff`, `ndwi` and `tide_m` which are 3D but with `time` and `quantile` dims), I worry that we might run into issues with efficiently indexing and selecting data downstream

I think a more Pythonic (["explicit is better than implicit"](https://peps.python.org/pep-0020/)) approach would be to only use `ds` to contain the actual final outputs of each layer, and pass any of the additional temporary inputs directly as `xr.DataArrays` to new parameters in each function. For example, rather than passing in `ds` that contains bands `ndwi`, `tide_m`, `freq_corr` to a function like this:
```
sample_func(ds)
```

Instead do this, so that all of the individual params can then be nicely documented in each function so users know exactly what they need to make each function work:
```
sample_func(
    ndwi=ds.ndwi,
    tide_m=ds.tide_m,
    freq_cor=ds.freq_cor)
```

And then rather than assigning the new outputs directly into `ds` within the functions like this:
```
def sample_func(ds):
    ...do stuff....
    ds["new_array"] = new_array
    return ds
    
ds = sample_func(...)    
```

Return the new array itself, and assign it into the master `ds` output of the function. That way you can also document exactly what the function outputs in the docstring, rather than having to describe the entire `ds`:

```
def sample_func(ds):
    ...do stuff....
    return new_array
    
ds["new_array"] = sample_func(...)        
```

So a sample workflow could be something like:
```
# Compute elevation and and return as master `ds`
ds, temp_layer1, temp_layer2 = elevation(...)

# Calculate extents and add to master `ds`
extents, temp_layer3 = extents(ds.elevation, temp_layer1, temp_layer2)
ds['extents'] = extents

# Calculate exposure and add to master `ds`
ds['exposure'] = pixel_exp(ds.extents, temp_layer3)

# Calculate spread and offset and add to master `ds`
ds['spread'], ds['offset'] = bias_offset(ds.extents)
```

File: intertidal/tidal_bias_offset.py [hunk 1]
```diff
@@ -1,40 +1,55 @@
 import xarray as xr
+import numpy as np
 
 def bias_offset(
                 ds,
-                dem,
-                set_dtype = None,
+                # set_dtype = None,
                 LAT_HAT=True,
```
[the remark this answers is not in the document]
Comment (erialC-P) [reply intertidal/tidal_bias_offset.py:7]: yes we can! Change commited

File: intertidal/tidal_bias_offset.py [hunk 2]
```diff
@@ -49,36 +64,40 @@ def bias_offset(
...
-        ds['LOT'] = min_obs.where(ds.tide_m.min(dim='time')> -9999)
-        ds['HOT'] = max_obs.where(ds.tide_m.min(dim='time')> -9999)
+        # ds['LAT'] = min_mod.where(ds.tide_m.min(dim='time')> -9999)
+        # ds['HAT'] = max_mod.where(ds.tide_m.min(dim='time')> -9999)
+        ds['LAT'] = min_mod.where(ds.extents == 1)
+        ds['HAT'] = max_mod.where(ds.extents == 1)
+    
+    ## Add the lowest and highest sensor-observed tides
+    ## TODO: mask using intertidal extent layer instead
+    if LOT_HOT is not None:
```
[the remark this answers is not in the document]
Comment (erialC-P) [reply intertidal/tidal_bias_offset.py]: done :)

File: intertidal/tidal_bias_offset.py [hunk 3]
```diff
@@ -49,36 +64,40 @@ def bias_offset(
...
     if LAT_HAT is True:
-        ds['LAT'] = min_mod.where(ds.tide_m.min(dim='time')> -9999)
-        ds['HAT'] = max_mod.where(ds.tide_m.min(dim='time')> -9999)
-        
-    if LOT_HOT is True:
-        ds['LOT'] = min_obs.where(ds.tide_m.min(dim='time')> -9999)
-        ds['HOT'] = max_obs.where(ds.tide_m.min(dim='time')> -9999)
+        # ds['LAT'] = min_mod.where(ds.tide_m.min(dim='time')> -9999)
+        # ds['HAT'] = max_mod.where(ds.tide_m.min(dim='time')> -9999)
+        ds['LAT'] = min_mod.where(ds.extents == 1)
```
[the remark this answers is not in the document]
Comment (erialC-P) [reply intertidal/tidal_bias_offset.py]: I think that was a late night coding mistake. I've bought back the always_dry only mask :)

File: intertidal/tidal_bias_offset.py [hunk 4]
```diff
@@ -1,40 +1,55 @@
...
+import numpy as np
 
 def bias_offset(
                 ds,
-                dem,
-                set_dtype = None,
+                # set_dtype = None,
                 LAT_HAT=True,
-                LOT_HOT=False
+                LOT_HOT=None
```
[the remark this answers is not in the document]
Comment (erialC-P) [reply intertidal/tidal_bias_offset.py]: changed :)

File: intertidal/exposure.py [hunk 5]
```diff
@@ -4,40 +4,58 @@
 from dea_tools.coastal import pixel_tides
 
 
-def pixel_exp(dem,
+def pixel_exp(ds,
               timerange, 
               directory='~/dev_intexp/dea-notebooks/tide_models_clipped'
```
[the remark this answers is not in the document]
Comment (erialC-P) [reply intertidal/exposure.py:9]: I forgot to do that. Good pick up!

**Comment by erialC-P** [comment]
Hey @robbibt, thanks for reviewing this PR. I agree with all your comments and the changes have considerably improved the work. Would you mind having a look through the revised workflow please and add any comments? Thank-you :)

**Review by robbibt — APPROVED** [review]
This is awesome @erialC-P, exactly what I had in mind! I've pushed a few additional commits that make some assorted, mostly minor changes:

- Make `ds` contain only the final outputs and nothing else, and rename some temporary xr.Datasets to differentiate them from `ds` (e.g. changing `ds` inside `elevation` to `satellite_ds`)
- Updated the exposure func to not need a custom dataset created as an input, and to work on the `ds.dem` array directly even though it doesn't have a time dimension
- Return `tide_m` as an output from `elevation` so it can be used later on and doesn't have to be included in `ds` (this may also make it easier to use a version of that data before we apply cloud masking to it)
- Changing the input order of some params so that the most important inputs are up front (e.g. `ds['exposure'], tide_cq = pixel_exp(ds.dem, **all_timerange)` vs `tide_cq, ds['exposure'] = pixel_exp(ds.dem, all_timerange)`)
- Some minor formatting updates and fixes to the notebook (e.g. getting the image export working, exporting data directly from `ds`

Have a look and check you're OK with the changes - happy to merge this as soon as you are!

**Pull request merged**